### ISSUE TITLE
Current master fails to build AVX2 Dilithium4 on CentOS 8

### ISSUE THREAD

**Issue by mouse07410** [PQClean/PQClean#359]
CentOS 8, GCC-8.3.1, current master (or rather, source directories taken and placed into the source tree of my project).

Basically, linker fails, complaining about `invntt.S.o` requiring `-fPIC` flag, despite that this flag was given. Here's the error:
```
[ 91%] Linking C shared library libmorf_crypto2.so
/usr/bin/cmake -E cmake_link_script CMakeFiles/morf_crypto2.dir/link.txt --verbose=1
/usr/bin/cc -fPIC -DMORF_KEYS_ARE_RED -DPQCLEAN_NTRUHPS4096821_AIX2_="" -DPQCLEAN_NTRUHPS4096821_CLEAN_="" -DPQCLEAN_DILITHIUM4_CLEAN_="" -DPQCLEAN_DILITHIUM4_AVX2_="" -DPQCLEAN_FALCON1024_AVX2_="" -DPQCLEAN_FALCON1024_CLEAN_=""  -Wall -O3 -O3 -march=native -std=gnu11 -DAVX2 -DKEMSPACE="PQCLEAN_NTRUHPS4096821_AVX2" -DSIGNSPACE="PQCLEAN_DILITHIUM4_AVX2"  -shared -Wl,-soname,libmorf_crypto2.so -o libmorf_crypto2.so CMakeFiles/morf_crypto2.dir/src/prng.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/cmov.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/crypto_sort_int32.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/kem.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/owcpa.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/pack3.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/packq.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_lift.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_r2_inv.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_s3_inv.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/sample.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/sample_iid.c.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_mod_3_Phi_n.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_mod_q_Phi_n.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_r2_mul.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_rq_mul.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/poly_rq_to_s3.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_102_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_12_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_1_821_patience.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_204_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_24_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_3_821_patience.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_408_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_51_821_shufbytes.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/square_6_821_patience.s.o ntruhps4096821/avx2/CMakeFiles/ntru.dir/vec32_sample_iid.s.o dilithium4/avx2/CMakeFiles/dilithium4.dir/fips202x4.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/nttconsts.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/packing.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/poly.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/polyvec.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/rejsample.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/rounding.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/sign.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/stream.c.o dilithium4/avx2/CMakeFiles/dilithium4.dir/invntt.S.o dilithium4/avx2/CMakeFiles/dilithium4.dir/ntt.S.o dilithium4/avx2/CMakeFiles/dilithium4.dir/pointwise.S.o dilithium4/avx2/CMakeFiles/dilithium4.dir/reduce.S.o pq_common/CMakeFiles/pqcommon.dir/fips202.c.o pq_common/CMakeFiles/pqcommon.dir/randombytes.c.o keccak4x/CMakeFiles/keccak4x.dir/KeccakP-1600-times4-SIMD256.c.o /opt/local/lib/libwhatever.a 
/usr/bin/ld: dilithium4/avx2/CMakeFiles/dilithium4.dir/invntt.S.o: relocation R_X86_64_PC32 against symbol `PQCLEAN_DILITHIUM4_AVX2_8xqinv' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

Here you can see that the source file in question indeed *was* compiled with `-fPIC` flag, which did not help:
```
.  .  .  .  .
[ 20%] Building ASM object dilithium4/avx2/CMakeFiles/dilithium4.dir/invntt.S.o
cd /home/ur20980/src/morf_crypto2/build/dilithium4/avx2 && /usr/bin/cc  -I/home/ur20980/src/morf_crypto2 -I/home/ur20980/src/morf_crypto2/src -I/home/ur20980/src/morf_crypto2/dilithium4/avx2 -I/home/ur20980/src/morf_crypto2/pq_common -I/home/ur20980/src/morf_crypto2/keccak4x  -O3 -fPIC -mavx2   -o CMakeFiles/dilithium4.dir/invntt.S.o -c /home/ur20980/src/morf_crypto2/dilithium4/avx2/invntt.S
[ 22%] Building ASM object dilithium4/avx2/CMakeFiles/dilithium4.dir/ntt.S.o
cd /home/ur20980/src/morf_crypto2/build/dilithium4/avx2 && /usr/bin/cc  -I/home/ur20980/src/morf_crypto2 -I/home/ur20980/src/morf_crypto2/src -I/home/ur20980/src/morf_crypto2/dilithium4/avx2 -I/home/ur20980/src/morf_crypto2/pq_common -I/home/ur20980/src/morf_crypto2/keccak4x  -O3 -fPIC -mavx2   -o CMakeFiles/dilithium4.dir/ntt.S.o -c /home/ur20980/src/morf_crypto2/dilithium4/avx2/ntt.S
[ 24%] Building ASM object dilithium4/avx2/CMakeFiles/dilithium4.dir/pointwise.S.o
.  .  .  .  .
```

Help is appreciated.

In case the complete log helps, here it is: [dilithium-out.txt](https://github.com/PQClean/PQClean/files/5564066/dilithium-out.txt)

### Update
I have the same problem on Ubuntu-20.04.1.

### Update 2
It looks like on MacOS X this file compiles OK because Clang-10 can properly parse it and process assembly commands. On Linux, GCC fails on assembly stuff, and YASM (that could process assembly) gets hung up on `cdecl()` macro, which it can't deal with.

### Update 3
I don't know, but I suspect that the difference between what I'm doing and what your CI does is that I'm adding your code to a *shared* library, and your CI and tests seem to be only building static libs, not validating that it integrates fine into dynamic libs.

**Comment by mouse07410** [comment]
Ping?

This is a bad one, pretty much preventing us from using optimized Dilithium...

**Comment by thomwiggers** [comment]
I've not got time to look into this; but I wonder why OQS would not be having trouble — they also use CMake and I think they also use shared libraries. 

If you have suggestions for improving CI we could review them, but unfortunately I also can't put in time at the moment to look into those things.

**Comment by mouse07410** [comment]
>  I wonder why OQS would not be having trouble — they also use CMake and I think they also use shared libraries.

Interesting. I rather doubt CMake is a factor here - but I did build successfully `liboqs` as a shared library (had to switch to static because OpenSSL tests consistently failed to locate that dynamic library!). 
Will investigate that.

> If you have suggestions for improving CI we could review them, but unfortunately I also can't put in time at the moment to look into those things

CI is not my strong point, but I'll take a look a bit later.

**Comment by jschanck** [comment]
OQS pulls dilithium directly from upstream (despite the pqclean_ folder name). See https://github.com/open-quantum-safe/liboqs/pull/794. This bug will be fixed by incorporating the new dilithium code into PQClean. If I have some free time I'll write a packaging script, but this is low priority as we're waiting for upstream to release the round 3 code.

**Comment by mouse07410** [comment]
> This bug will be fixed by incorporating the new dilithium code into PQClean . . .

Thanks!

> . . .we're waiting for upstream to release the round 3 code. . . 

Could you share any ballpark-level "when"? Thanks!

**Comment by jschanck** [comment]
@mouse07410 
Round 3 Dilithium is now available (commit 4f86c3951562af84ea4036a7e0483a9e96cebba9).
